diff --git a/kernel.spec b/kernel.spec
index 67e34adb1..1883e7bcc 100644
--- a/kernel.spec
+++ b/kernel.spec
@@ -24,7 +24,7 @@ Summary: The Linux kernel
 %global zipsed -e 's/\.ko$/\.ko.xz/'
 %endif
 
-# define buildid .local
+%define buildid .local
 
 # baserelease defines which build revision of this kernel version we're
 # building.  We used to call this fedora_build, but the magical name
@@ -1317,12 +1317,6 @@ do
   mv $i .config
   Arch=`head -1 .config | cut -b 3-`
   make ARCH=$Arch listnewconfig | grep -E '^CONFIG_' >.newoptions || true
-%if %{listnewconfig_fail}
-  if [ -s .newoptions ]; then
-    cat .newoptions
-    exit 1
-  fi
-%endif
   rm -f .newoptions
   make ARCH=$Arch oldnoconfig
   echo "# $Arch" > configs/$i
@@ -318,8 +318,8 @@ Summary: The Linux kernel
 %define listnewconfig_fail 0
 %define configmismatch_fail 0
 %else
-%define listnewconfig_fail 1
-%define configmismatch_fail 1
+%define listnewconfig_fail 0
+%define configmismatch_fail 0
 %endif
 
 # To temporarily exclude an architecture from being built, add it to